go/enclave/crosschain/obscuro_message_bus_manager.go

```diff
@@ -0,0 +1,254 @@
+	sender, err := rpc.GetSender(&transaction)
+	if err != nil {
+		return false
+	}
+	// The message bus manager considers the transaction synthetic only if the sender is
+	// the owner identity which should be available only to enclaves.
+	return bytes.Equal(sender.Bytes(), m.GetOwner().Bytes())
+}
+
+// GetOwner - Returns the address of the identity owning the message bus.
+func (m *obscuroMessageBusManager) GetOwner() common.L2Address {
+	return m.wallet.Address()
+}
+
+// GetBusAddress - Returns the L2 address of the message bus contract.
+// TODO: Figure out how to expose the deployed contract to the external world. Perhaps extract event from contract construction?
+func (m *obscuroMessageBusManager) GetBusAddress() *common.L2Address {
+	return m.messageBusAddress
+}
+
+// DeriveOwner - Generates the key pair that will be used to transact with the L2 message bus.
+func (m *obscuroMessageBusManager) DeriveOwner(seed []byte) (*common.L2Address, error) {
+	// TODO: Implement with cryptography epic!
+	return m.messageBusAddress, nil
+}
+
+// GenerateMessageBusDeployTx - Returns a signed message bus deployment transaction.
+func (m *obscuroMessageBusManager) GenerateMessageBusDeployTx() (*common.L2Tx, error) {
+	tx := &types.LegacyTx{
+		Nonce:    0, // The first transaction of the owner identity should always be deploying the contract
+		Value:    gethcommon.Big0,
+		Gas:      5_000_000,       // It's quite the expensive contract.
+		GasPrice: gethcommon.Big0, // Synthetic transactions are on the house. Or the house.
+		Data:     gethcommon.FromHex(MessageBus.MessageBusMetaData.Bin),
+		To:       nil, // Geth requires nil instead of gethcommon.Address{} which equates to zero address in order to return receipt.
+	}
+
+	stx, err := m.wallet.SignTransaction(tx)
+	if err != nil {
+		return nil, err
+	}
+
+	m.logger.Trace(fmt.Sprintf("[CrossChain] Generated synthetic deployment transaction for the MessageBus contract %s - TX HASH: %s", m.messageBusAddress.Hex(), stx.Hash().Hex()))
+
+	return stx, nil
+}
+
+// ExtractLocalMessages - Finds relevant logs in the receipts and converts them to cross chain messages.
+func (m *obscuroMessageBusManager) ExtractLocalMessages(receipts common.L2Receipts) (common.CrossChainMessages, error) {
+	logs, err := filterLogsFromReceipts(receipts, m.messageBusAddress, &CrossChainEventID)
+	if err != nil {
+		m.logger.Error("[CrossChain] Error extracting logs from L2 message bus!", "Error", err)
+		return make(common.CrossChainMessages, 0), err
+	}
+
+	messages, err := convertLogsToMessages(logs, CrossChainEventName, MessageBusABI)
+	if err != nil {
+		m.logger.Error("[CrossChain] Error converting messages from L2 message bus!", "Error", err)
+		return make(common.CrossChainMessages, 0), err
+	}
+
+	return messages, nil
+}
+
+// SubmitRemoteMessagesLocally - Submits messages saved between the from and to blocks on chain using the provided function bindings.
+func (m *obscuroMessageBusManager) SubmitRemoteMessagesLocally(
```

Comment:
I tried to get my head around this method and it feels like it's doing too much.
Also not a fan of those two wrapper functions which invert the logic .

How about in rollupchain. process state you do the following:
```
// these are the 2 methods splt from retrieveSyntheticTransactionsBetween
   messages := rc.crossChainProcessors.LocalManager.RetrieveCrosschainMessages(from, to..)
   syntheticTxs := rc.crossChainProcessors.LocalManager.CreateSynthTx(messages,...)
  evmRep:=  evm.ExecuteTransactions(syntheticTxs, stateDB, rollup.Header, rc.storage, rc.chainConfig, len(executedTransactions), rc.logger) 
.... receipts etc
```

This way the entire action is more explicit and linear.

Reply:
I had it initially like this, but it felt as if the rollup chain had too much of a responsibility and was doing everything. I can see how the function wrappers might not be the right solution to that problem tho.